### PR TITLE
Scarcity Manifested ERC721 upgraded

### PR DESCRIPTION
**Pull request by eirlis**

### CODE REVIEW

File: contracts/core/rarity.sol [hunk 1]
```diff
@@ -59,77 +103,391 @@ interface IERC721Receiver {
...
+
+    function _msgData() internal view virtual returns (bytes calldata) {
+        return msg.data;
+    }
+}
+
+/**
+ * @dev Collection of functions related to the address type
+ */
+library Address {
```
Comment (ytumanov) [inline contracts/core/rarity.sol:140]: can be imported from openzeppelin

File: contracts/core/rarity.sol [hunk 2]
```diff
@@ -1,7 +1,51 @@
...
+ *
+ * ```solidity
+ * function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+ *     return interfaceId == type(MyInterface).interfaceId || super.supportsInterface(interfaceId);
+ * }
+ * ```
+ *
+ * Alternatively, {ERC165Storage} provides an easier to use but more expensive implementation.
+ */
+abstract contract ERC165 is IERC165 {
```
Comment (ytumanov) [inline contracts/core/rarity.sol:39]: can be imported from openzeppelin and other contracts

File: contracts/core/rarity.sol [hunk 3]
```diff
@@ -561,8 +921,8 @@ contract rarity is ERC721Enumerable {
     uint constant xp_per_day = 250e18;
     uint constant DAY = 1 days;
 
-    string constant public name = "Scarcity Manifested";
-    string constant public symbol = "SM";
+    // string constant public name = "Scarcity Manifested";
```
Comment (ytumanov) [inline contracts/core/rarity.sol:924]: remove?

File: contracts/core/rarity.sol [hunk 4]
```diff
@@ -571,6 +931,10 @@ contract rarity is ERC721Enumerable {
 
     event summoned(address indexed owner, uint class, uint summoner);
     event leveled(address indexed owner, uint level, uint summoner);
+    
+    constructor() ERC721("Scarcity Manifested", "SM") {
+        
+    }
 
     function adventure(uint _summoner) external {
         require(_isApprovedOrOwner(msg.sender, _summoner));
```
Comment (ytumanov) [inline contracts/core/rarity.sol:940]: _msgSender()

**Review by ytumanov — CHANGES_REQUESTED** [review]

**Review by hayalet27 — APPROVED** [review]

**Review by hayalet27 — APPROVED** [review]

**Review by hayalet27 — CHANGES_REQUESTED** [review]
Kontrol edilsin

**Review by ytumanov — APPROVED** [review]

**Pull request merged**